refactor api

elsabagh opened the pull request.

elsabagh committed:
Handle validation and solve problems with name country

elsabagh committed:
Merge remote-tracking branch 'origin/develop' into feature/handling-validation

elsabagh committed:
Handle validation in ui

elsabagh committed:
refactor location api

yusufnasserdev commented on src/main/kotlin/presentation/ui/ClothesSuggesterByCityNameCli.kt:
```diff
@@ -70,4 +74,20 @@ class ClothesSuggesterByCityNameCli(
         printer.displayLn("\n🕺 Go rule the streets. Or at least your hallway.".withStyle(TerminalColor.Magenta))
     }
+
+    private fun readValidInput(prompt: String): String {
+        while (true) {
+            val input = inputReader.readString(prompt)
+
+            if (!isValidName(input)) {
+                printer.displayLn("Invalid input! Please enter letters only.".withStyle(TerminalColor.Red))
+            } else {
+                return input
+            }
+        }
+    }
```
Why can't this be a part of the input reader implementation? also why not implement the validation in the already existing `readString` function.

elsabagh replied:
done

elsabagh committed:
create a search by city and create a test case

- Create a search by city to suggest a listing that matches your city name in many different countries.

yusufnasserdev commented on src/main/kotlin/domain/usecase/SearchCityUseCase.kt:
```diff
@@ -0,0 +1,12 @@
+class SearchCityUseCase(
+    private val locationRepository: LocationRepository
+) {
+    suspend operator fun invoke(city: String): List<LocationDto> {
```
A `Dto` shouldn't be used on the domain package, especially in the use-case.

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/data/weather/datasource/WeatherApiClient.kt:
```diff
@@ -28,6 +28,10 @@ class WeatherApiClient(
         return getWeatherByLocation(ipLocation.toLocationDto())
     }
 
+    override suspend fun getWeatherBySearch(location: LocationDto): WeatherDto {
+        return getWeatherByLocation(location)
+    }
+
```
What's the point of this function, it only calls another existing one?

elsabagh replied:
After the required changes, it was removed and GetWeatherUseCase was modified.

yusufnasserdev commented on src/main/kotlin/domain/usecase/GetWeatherBySearchUseCase.kt:
```diff
@@ -0,0 +1,13 @@
+import org.damascus.domain.model.WeatherInfo
+import org.damascus.domain.repository.WeatherRepository
+
+class GetWeatherBySearchUseCase(
```
This name is vague, what are you searching for?

elsabagh replied:
After the required changes, it was removed because now we use GetWeatherUseCase.

elsabagh committed:
merge duplicate files, create a location mapper, and modify a test case
